src/contracts/markets/SiloMarket.sol

```diff
@@ -0,0 +1,192 @@
+        return IERC4626(market).balanceOf(address(this));
+    }
+
+    /// @notice Deposit a exact amount of asset tokens to the Silo lending market
+    /// and mint a variable amount of Silo lending market shares to this contract.
+    /// @param assets The exact amount of asset tokens to deposit.
+    /// @param receiver The receiver has to be the address of the ARM contract.
+    /// @return shares The amount of Silo lending market shares that were minted.
+    function deposit(uint256 assets, address receiver) external returns (uint256 shares) {
+        require(msg.sender == arm && receiver == arm, "Only ARM can deposit");
+
+        // Transfer liquidity from the ARM to this contract
+        IERC20(asset).transferFrom(arm, address(this), assets);
+
+        // Approve the Silo lending market to spend the asset tokens
+        IERC20(asset).approve(market, assets);
+        // Deposit assets to the Silo lending market from this contract
+        // and mint shares to this contract
+        shares = IERC4626(market).deposit(assets, address(this));
+
+        emit Deposit(arm, arm, assets, shares);
+    }
+
+    /// @notice Get the max amount of asset tokens that can be withdrawn from the Silo lending market
+    /// from the Silo lending market shares owned by this contract.
+    /// @param owner The owner account has to be the address of the ARM contract.
+    /// @return maxAssets The max amount of asset tokens that can be withdrawn.
+    function maxWithdraw(address owner) external view returns (uint256 maxAssets) {
+        if (owner != arm) return 0;
+
+        maxAssets = IERC4626(market).maxWithdraw(address(this));
+    }
+
+    /// @notice Withdraw a exact amount of asset tokens from the Silo lending market
+    /// from the Silo lending market shares owned by this contract.
+    /// @param assets The exact amount of asset tokens to withdraw.
+    /// @param receiver The receiver has to be the address of the ARM contract.
+    /// @param owner The owner has to be the address of the ARM contract.
+    /// @return shares The amount of Silo lending market shares that were burnt.
+    function withdraw(uint256 assets, address receiver, address owner) external returns (uint256 shares) {
+        require(msg.sender == arm && receiver == arm && owner == arm, "Only ARM can withdraw");
+
+        // Withdraw assets from the lending market to the ARM
+        shares = IERC4626(market).withdraw(assets, arm, address(this));
+
+        emit Withdraw(arm, arm, arm, assets, shares);
+    }
+
+    /// @notice Get the amount of asset tokens that can be received
+    /// from burning an exact amount of Silo lending market shares.
+    /// @return assets The amount of asset tokens that will be received.
+    function previewRedeem(uint256 shares) external view returns (uint256 assets) {
+        // Preview the amount of assets that can be redeemed for a given number of shares
+        assets = IERC4626(market).previewRedeem(shares);
```

Comment:
I wonder if it's worth checking that reentrancyGuardEntered is false, so that we don't have to worry about any kind of read only reentrancy.

![image](https://github.com/user-attachments/assets/8349453c-88c2-4946-aab8-86895b8a9be7)

Reply:
Bumping this. Silo is crawling with hooks. In theory they've got everything nicely covered and correctly ordered, but I would much rather be checking this for our silo integration.